### PR TITLE
gormschema: fixed circular foreign keys on postgresql example

### PR DESCRIPTION
**Pull request by dorav**
Fixes Issue #22

### CODE REVIEW

**Comment by rotemtam** [comment]
CI is red @dorav

File: gormschema/gorm.go [hunk 1]
```diff
@@ -73,12 +74,79 @@ func (l *Loader) Load(models ...any) (string, error) {
...
+	gorm.Dialector
+}
+
+func (d customDialector) newCustomMigrator(db *gorm.DB) *customMigrator {
+	return &customMigrator{
+		Migrator: migrator.Migrator{
+			Config: migrator.Config{
+				DB:                          db,
+				Dialector:                   d,
+				CreateIndexAfterCreateTable: true,
```
Comment (rotemtam) [inline gormschema/gorm.go:116]: Why is this necessary?

File: gormschema/gorm.go [hunk 2]
```diff
@@ -73,12 +74,79 @@ func (l *Loader) Load(models ...any) (string, error) {
...
 	}
 	s, ok := recordriver.Session("gorm")
 	if !ok {
-		return "", err
+		return "", fmt.Errorf("gorm db session not found")
 	}
 	return s.Stmts(), nil
 }
+
+type customMigrator struct {
```
Comment (rotemtam) [inline gormschema/gorm.go:101]:
```suggestion
type migrator struct {
```

File: gormschema/gorm.go [hunk 3]
```diff
@@ -73,12 +74,79 @@ func (l *Loader) Load(models ...any) (string, error) {
...
+	migrator.Migrator
+	dialectMigrator gorm.Migrator
+}
+
+type customDialector struct {
```
Comment (rotemtam) [inline gormschema/gorm.go:106]:
```suggestion
type dialector struct {
```

File: gormschema/gorm.go [hunk 4]
```diff
@@ -73,12 +74,79 @@ func (l *Loader) Load(models ...any) (string, error) {
...
+
+func (d customDialector) Migrator(db *gorm.DB) gorm.Migrator {
+	return d.newCustomMigrator(db)
+}
+
+func (m *customMigrator) HasTable(dst interface{}) bool {
+	return true
+}
+
+func (m *customMigrator) CreateConstraints(models []interface{}) error {
```
Comment (rotemtam) [inline gormschema/gorm.go:131]: add godoc for all public methods, especially this one since it contains the special logic

File: gormschema/gorm_test.go [hunk 5]
```diff
@@ -20,3 +21,29 @@ func TestConfig(t *testing.T) {
 	require.Contains(t, sql, "CREATE TABLE `users`")
 	require.NotContains(t, sql, "FOREIGN KEY")
 }
+
+func TestPostgreSQLConfig(t *testing.T) {
```
Comment (rotemtam) [inline gormschema/gorm_test.go:25]: add test for mysql + sqlite

File: gormschema/gorm.go [hunk 6]
```diff
@@ -73,12 +74,79 @@ func (l *Loader) Load(models ...any) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	if err := db.AutoMigrate(models...); err != nil {
-		return "", err
+	db.Config.DisableForeignKeyConstraintWhenMigrating = true
+	err = db.AutoMigrate(models...)
+	if !l.config.DisableForeignKeyConstraintWhenMigrating {
```
Comment (rotemtam) [inline gormschema/gorm.go:79]: is it possible to invoke the new logic only if cycles are detected?

**Review by rotemtam — COMMENTED** [review]
Well done !! 



Some suggestions and questions

File: gormschema/gorm.go (hunk 6 above)
Comment (dorav) [reply gormschema/gorm.go:81]: Not that I'm aware of, I can't find any clear error from AutoMigrate that we can use. 
We could probably search the graph, but that would require some more research

File: gormschema/gorm_test.go (hunk 5 above)
Comment (dorav) [reply gormschema/gorm_test.go:25]: sqlite does not support this use-case

File: gormschema/gorm.go [hunk 7]
```diff
@@ -10,6 +10,7 @@ import (
 	"gorm.io/driver/postgres"
 	"gorm.io/driver/sqlite"
 	"gorm.io/gorm"
+	gormMig "gorm.io/gorm/migrator"
```
Comment (a8m) [inline gormschema/gorm.go:13]:
```suggestion
	gormigrate "gorm.io/gorm/migrator"
```

File: gormschema/gorm.go [hunk 8]
```diff
@@ -24,8 +25,9 @@ func New(dialect string, opts ...Option) *Loader {
 type (
 	// Loader is a Loader for gorm schema.
 	Loader struct {
-		dialect string
-		config  *gorm.Config
+		dialect                           string
+		createConstraintsAfterCreateTable bool
```
Comment (a8m) [inline gormschema/gorm.go:29]: change it to `detachFKs`/`deferredFKs` - anything that looks like the current naming

File: gormschema/gorm.go [hunk 9]
```diff
@@ -38,6 +40,13 @@ func WithConfig(cfg *gorm.Config) Option {
 	}
 }
 
+// WithCreateConstraintsAfterCreateTable sets the createConstraintsAfterCreateTable config.
+func WithCreateConstraintsAfterCreateTable(b bool) Option {
+	return func(l *Loader) {
+		l.createConstraintsAfterCreateTable = b
+	}
+}
```
Comment (a8m) [inline gormschema/gorm.go:48]: ditto. It does not look like Atlas API. 


Also, why the user should be aware of this? This is encapsulated by Atlas and should be different here

Comment (dorav) [reply gormschema/gorm.go]: I'll remove it

File: gormschema/gorm_test.go [hunk 10]
```diff
@@ -3,20 +3,105 @@ package gormschema
...
+	l := New("sqlite")
+	sql, err := l.Load(models.Pet{}, models.User{}, ckModels.Event{}, ckModels.Location{})
+	require.NoError(t, err)
+	require.Contains(t, sql, "CREATE TABLE `pets`")
+	require.Contains(t, sql, "CREATE TABLE `users`")
+	require.Contains(t, sql, "CREATE TABLE `events`")
+	require.Contains(t, sql, "CREATE TABLE `locations`")
+	require.Contains(t, sql, "FOREIGN KEY (`eventId`)")
+	require.Contains(t, sql, "FOREIGN KEY (`locationId`)")
+	require.Contains(t, sql, "CONSTRAINT `fk_users_pets` FOREIGN KEY")
```
Comment (a8m) [inline gormschema/gorm_test.go:23]: Use equal comparison (and not contains) as the order of the statements does matter.

File: gormschema/gorm_test.go [hunk 11]
```diff
@@ -3,20 +3,105 @@ package gormschema
...
+	resetSession(t)
+	l = New("postgres",
+		WithCreateConstraintsAfterCreateTable(true))
+	sql, err = l.Load(ckModels.Location{}, ckModels.Event{})
+	require.NoError(t, err)
+	require.Contains(t, sql, `CREATE TABLE "events"`)
+	require.Contains(t, sql, `CREATE UNIQUE INDEX IF NOT EXISTS "idx_events_location_id"`)
+	require.Contains(t, sql, `CREATE TABLE "locations"`)
+	require.Contains(t, sql, `CREATE UNIQUE INDEX IF NOT EXISTS "idx_locations_event_id"`)
+	require.Contains(t, sql, `ALTER TABLE "events" ADD CONSTRAINT "fk_locations_event" FOREIGN KEY ("locationId")`)
```
Comment (a8m) [inline gormschema/gorm_test.go:59]: ditto

File: internal/testdata/circular_fk_models/models.go [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package circular_fk_models
```
Comment (a8m) [inline internal/testdata/circular_fk_models/models.go:1]:
```suggestion
package circularfks
```
don't use _ in package names.

File: gormschema/gorm_test.go [hunk 13]
```diff
@@ -3,20 +3,105 @@ package gormschema
...
-	require.Contains(t, sql, "CREATE TABLE `users`")
+	require.Contains(t, sql, "CREATE TABLE `events`")
+	require.Contains(t, sql, "CREATE TABLE `locations`")
 	require.NotContains(t, sql, "FOREIGN KEY")
 }
+
+func resetSession(t *testing.T) {
+	sess, ok := recordriver.Session("gorm")
+	require.True(t, ok)
+	sess.Statements = []string{}
```
Comment (a8m) [inline gormschema/gorm_test.go:106]:
```suggestion
	sess.Statements = nil
```

File: gormschema/gorm.go [hunk 14]
```diff
@@ -73,12 +74,82 @@ func (l *Loader) Load(models ...any) (string, error) {
...
+			return "", err
+		}
+		if err = cm.CreateConstraints(models); err != nil {
+			return "", err
+		}
+	}
 	s, ok := recordriver.Session("gorm")
 	if !ok {
-		return "", err
+		return "", fmt.Errorf("gorm db session not found")
```
Comment (a8m) [inline gormschema/gorm.go:100]: use errors.New

**Review by a8m — APPROVED** [review]
